### PR TITLE
Spiro octahedrals

### PR DESCRIPTION
**Pull request by johnmay**
Previously I would just refuse to lay out geometries where we have a ring system embedded, now we can handle some cases:

image

Will do a release once this is in, always more to do but it's time I think

### CODE REVIEW

**Comment by johnmay** [comment]
Oh here are the SMILES:

```
O=C1O[Fe@](OC1=O)(OC2=O)(Cl)OC2=O
O=C1O[Fe@@](OC1=O)(OC2=O)(Cl)OC2=O
O=C1O[Fe@OH12-3]23(OC(C(O3)=O)=O)(OC(C(O2)=O)=O)OC1=O
O=C1O[Fe@OH13-3]23(OC(C(O3)=O)=O)(OC(C(O2)=O)=O)OC1=O
Cl[Fe@OH12-3]23(OC(C(O3)=O)=O)(OC(C(O2)=O)=O)Cl
Cl[Fe@OH13-3]23(OC(C(O3)=O)=O)(OC(C(O2)=O)=O)Cl
```

File: base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java [hunk 1]
```diff
@@ -254,6 +255,39 @@ public static void rotate(IAtomContainer atomCon, Point2d center, double angle)
...
     }
 
+    /**
+     * Reflect a collection of atoms in the line formed by the two specified points (beg,end).
+     *
+     * @param atoms the atoms
+     * @param beg the begin point of a line
+     * @param end the end point of a line
+     */
+    public static void reflect(Collection<IAtom> atoms, Point2d beg, Point2d end) {
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java:265]: My first thought was,  "but that changes the stereochemistry..." But then I realized that we are no longer using concepts like "wedge bond up" as model of stereochemistry :)

Comment (egonw) [reply base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java:265]: Can you please add a unit test for it?

File: base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java [hunk 2]
```diff
@@ -254,6 +255,39 @@ public static void rotate(IAtomContainer atomCon, Point2d center, double angle)
...
+        }
+    }
+
+    /**
+     * Reflect a collection of atoms in the line formed by the specified bond.
+     *
+     * @param atoms the atoms
+     * @param bond the bond at which to reflect
+     */
+    public static void reflect(Collection<IAtom> atoms, IBond bond) {
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java:286]: Can you please add a unit test for it?

File: base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java [hunk 3]
```diff
@@ -1683,22 +1717,30 @@ public static double getBondLengthAverage(IReaction reaction) {
...
     public static double getBondLengthMedian(final IAtomContainer container) {
-        if (container.getBondCount() == 0) throw new IllegalArgumentException("Container has no bonds.");
-        int nBonds = 0;
-        double[] lengths = new double[container.getBondCount()];
-        for (int i = 0; i < container.getBondCount(); i++) {
-            final IBond bond = container.getBond(i);
+        return getBondLengthMedian(container.bonds(), container.getBondCount());
+    }
+
+    public static double getBondLengthMedian(final Iterable<IBond> bonds, int cap) {
```
Comment (egonw) [inline base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java:1723]: Please add Javadoc.

**Review by egonw — CHANGES_REQUESTED** [review]
Two minor requests. Great patch!

File: base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java (hunk 1 above)
Comment (johnmay) [reply base/standard/src/main/java/org/openscience/cdk/geometry/GeometryUtil.java:265]: You are right, it should flip bonds as meant to be a generic method

**Comment by johnmay** [comment]
Pushed an update

**Comment by egonw** [comment]
> Pushed an update

@johnmay, good to go? (You can click the "recycle"-like icon to indicate I should look at it again, and with that, that you do not plan additional updates...)

**Pull request merged**